src/librustc_const_eval/eval.rs

```diff
@@ -289,6 +267,7 @@ fn eval_const_expr_partial<'a, 'tcx>(cx: &ConstContext<'a, 'tcx>,
           match cx.tables.qpath_def(qpath, e.hir_id) {
               Def::Const(def_id) |
               Def::AssociatedConst(def_id) => {
+                    let substs = tcx.normalize_associated_type_in_env(&substs, cx.param_env);
```

Comment:
Isn't this normalization also needed in the `Def::Method(id) | Def::Fn(id)` arm?

Reply:
technically we only need it for `Def::AssociatedConst`, but it's a nop for `Def::Const`